File: hash/consistency-hash-package/ConsistencyHash.cpp
#include "ConsistencyHash.h"

ConsistencyHash::ConsistencyHash(const std::vector<std::string>& nodes, const int& vnodes) {
    this->nodes = nodes;
    this->vnodes = vnodes;
    // printf("827 + 9912 = %" PRIu64 "\n", nodeHash(827, 9912));
    for(int i = 0; i < nodes.size(); i++) {
        for(int j = 0; j < vnodes; j++) {
            uint64_t h = nodeHash(i, j);
            this->ring.push_back(h);
            nodeByHash[h] = this->nodes[i];
        }
    }

    std::sort(this->ring.begin(), this->ring.end());
}

std::string ConsistencyHash::getNode(const std::string& data) {
    uint64_t h = hash(data.c_str(), data.size());
    if(h > ring.back()) {
        return nodes[0];
    }
    

    // 二分搜索优化
    int partition = bisectLeft(h);
    return nodeByHash[ring[partition]];

    // for(int i = 0; i < ringSize; i++) {
    //     if(h <= ring[i]) {
    //         std::unordered_map<uint64_t, std::string>::const_iterator iter = nodeByHash.find(ring[i]);
    //         if(iter != nodeByHash.end()) {
    //             return iter->second;
    //         }
    //         // return nodeByHash[ring[i]];
    //     }
    // }
}

int ConsistencyHash::bisectLeft(uint64_t value) {
    int l = 0, r = ring.size();
    while(l < r) {
        int mid = l + ((r - l) >> 1);
        if(ring[mid] < value) l = mid + 1;
        else r = mid;
    }
    return l;
}

uint64_t ConsistencyHash::hash(const char *src, int length) {
    unsigned char h[16];
    MD5_CTX ctx;
    MD5_Init(&ctx);
    MD5_Update(&ctx, src, length);
    MD5_Final(h, &ctx);

    // for(int i = 0; i < 16; i++) {
    //     printf("%02x", h[i]);
    // }
    uint64_t v = 0;
    for(int i = 0; i < 8; i++) 
        v |= ((uint64_t) h[i + 4]) << (56 - (i << 3));

    return v;
}

uint64_t ConsistencyHash::nodeHash(int i, int j) {
    char src[8];
    src[0] = i >> 24;
    src[1] = i >> 16;
    src[2] = i >> 8;
    src[3] = i;
    src[4] = j >> 24;
    src[5] = j >> 16;
    src[6] = j >> 8;
    src[7] = j;
    return hash(src, 8);
}

File: hash/consistency-hash-package/consistency_hash.py
# coding=utf-8
import hashlib
import struct
import bisect

class ConsitencyHash:
    def __init__(self, nodes, vnodes=5000):
        """初始化一致性哈希
        Args:
            nodes: 传入实际节点列表, 可以为代号, 比如: [1, 2, 3]
            vnodes: 每个节点对应虚拟节点个数, 默认: 5000 集群规模越大, vnodes可以越小
        """
        self.nodes = nodes

        self.nodes_count = len(self.nodes)
        self.vnodes_count = vnodes

        # hash环
        self.ring = []

        # 虚节点hash -> 实际节点映射
        self.node_by_hash = {}

        for i in range(self.nodes_count):
            for j in range(self.vnodes_count):
                src = struct.pack('>2I', i, j)
                h = self.__hash(src)
                self.ring.append(h)
                self.node_by_hash[h] = self.nodes[i]

        # 排序
        self.ring.sort()


    def __hash(self, src: bytes) -> int:
        """hash函数"""
        m = hashlib.md5()
        m.update(src)
        d = m.digest()[4:12]
        return struct.unpack_from('>L', d, 0)[0]

    def get_node(self, data: bytes):
        """给定数据返回node节点
        """
        h = self.__hash(data)

        # 计算插入区间 左开右闭 边界值插入到左边区间
        partition = bisect.bisect_left(self.ring, h)
        if partition == len(self.ring):
            return self.nodes[0]
        return self.node_by_hash[self.ring[partition]]
